### PR TITLE
RVC DirectModeChange Feature Update for RVCCLEANM and RVCRUNM

### PR DESCRIPTION
**Pull request by Harshith-GRL**
Added Check for DirectModeChange Feature when changing mode in RunMode and CleanMode.
Fix For Issue https://github.com/CHIP-Specifications/chip-test-plans/issues/4202
Scripts changed in this PR TC_RVCRUNM_2_2.py and TC_RVCCLEANM_2_2.py.
Fix for issue https://github.com/project-chip/connectedhomeip/issues/34946  
Update python Script TC_RVCOPSTATE_2_4.py

### CODE REVIEW

**Review by kiel-apple — APPROVED** [review]

**Review by kiel-apple — APPROVED** [review]

File: src/python_testing/TC_RVCRUNM_2_2.py [hunk 1]
```diff
@@ -38,17 +39,25 @@
 # --int-arg PIXIT.RVCRUNM.MODE_A:<mode id> PIXIT.RVCRUNM.MODE_B:<mode id>
 
 
+class RvcStatusEnum(enum.IntEnum):
+    # TODO remove this class once InvalidInMode response code is implemented in python SDK
```
Comment (rbultman) [inline src/python_testing/TC_RVCRUNM_2_2.py:43]: Please create an issue for this.

**Review by rbultman — CHANGES_REQUESTED** [review]

File: src/python_testing/TC_RVCCLEANM_2_2.py [hunk 2]
```diff
@@ -27,13 +27,22 @@
...
 from time import sleep
 
 import chip.clusters as Clusters
-from matter_testing_support import MatterBaseTest, async_test_body, default_matter_test_main
+from matter_testing_support import MatterBaseTest, async_test_body, default_matter_test_main, type_matches
 from mobly import asserts
 
 
+class RvcStatusEnum(enum.IntEnum):
+    # TODO remove this class once InvalidInMode response code is implemented in python SDK
```
[the remark this answers is not in the document]
Comment (kiel-apple) [reply src/python_testing/TC_RVCCLEANM_2_2.py:39]: #35031

File: src/python_testing/TC_RVCCLEANM_2_2.py [hunk 3]
```diff
@@ -90,6 +104,9 @@ def write_to_app_pipe(self, command):
 
     @async_test_body
     async def test_TC_RVCCLEANM_2_2(self):
+        # TODO Replace 0x8000 with python object of RVCCLEAN FEATURE bit map when implemented
+        # 0x8000 corresponds to 16 bit DIRECTMODECH Feature map
+        self.directmodech_bit_map = 0x8000
```
[the remark this answers is not in the document]
Comment (kiel-apple) [reply src/python_testing/TC_RVCCLEANM_2_2.py:109]: #35031

File: src/python_testing/TC_RVCCLEANM_2_2.py [hunk 4]
```diff
@@ -157,11 +174,25 @@ async def test_TC_RVCCLEANM_2_2(self):
...
-                             "with the Status set to InvalidInMode(0x03).")
+        self.print_step("7a", "Read FeatureMap Attribute")
+        directmodech = await self.read_feature_map_attribute()
+        directmode_enabled = directmodech & self.directmodech_bit_map
+
+        self.print_step("7b", "Send ChangeToMode command")
+        if directmode_enabled:
+            response = await self.send_clean_change_to_mode_cmd(self.new_clean_mode_th)
+            asserts.assert_true(type_matches(response, Clusters.RvcCleanMode.Commands.ChangeToModeResponse),
+                                "The response should ChangeToModeResponse command")
```
[the remark this answers is not in the document]
Comment (kiel-apple) [reply src/python_testing/TC_RVCCLEANM_2_2.py:185]: +1

File: src/python_testing/TC_RVCRUNM_2_2.py (hunk 1 above)
Comment (kiel-apple) [reply src/python_testing/TC_RVCRUNM_2_2.py:43]: I'm surprised this isn't already there.

File: src/python_testing/TC_RVCRUNM_2_2.py [hunk 5]
```diff
@@ -122,6 +131,9 @@ async def test_TC_RVCRUNM_2_2(self):
                          "PIXIT.RVCRUNM.MODE_A:<mode id> \n"
                          "PIXIT.RVCRUNM.MODE_B:<mode id>")
 
+        # TODO Replace 0x8000 with python object of RVCRUN FEATURE bit when implemented
```
[the remark this answers is not in the document]
Comment (kiel-apple) [reply src/python_testing/TC_RVCRUNM_2_2.py:134]: #35031

File: src/python_testing/TC_RVCRUNM_2_2.py (hunk 1 above)
Comment (Harshith-GRL) [reply src/python_testing/TC_RVCRUNM_2_2.py:43]: https://github.com/project-chip/connectedhomeip/issues/35031

File: src/python_testing/TC_RVCRUNM_2_2.py [hunk 6]
```diff
@@ -193,15 +205,23 @@ async def test_TC_RVCRUNM_2_2(self):
...
+        await self.send_change_to_mode_with_check(self.mode_a, RvcStatusEnum.Success)
         # This step is not described in the test plan, but it ought to be
         await self.read_current_mode_with_check(self.mode_a)
 
-        self.print_step(6, "Send ChangeToMode MODE_B command")
-        await self.send_change_to_mode_with_check(self.mode_b, 3)
+        self.print_step("6a", "Read Attribute FeatureMap")
+        directmodech = await self.read_mod_attribute_expect_success(cluster=Clusters.RvcRunMode,
+                                                                    attribute=Clusters.RvcRunMode.Attributes.FeatureMap)
+        directmode_enabled = directmodech & self.directmodech_bit_map
```
[the remark this answers is not in the document]
Comment (Harshith-GRL) [reply src/python_testing/TC_RVCRUNM_2_2.py]: Re-named variables as per comments

File: src/python_testing/TC_RVCCLEANM_2_2.py [hunk 7]
```diff
@@ -157,11 +174,25 @@ async def test_TC_RVCCLEANM_2_2(self):
...
                 break
 
-        self.print_step(7, "Send ChangeToMode command")
-        response = await self.send_clean_change_to_mode_cmd(self.new_clean_mode_th)
-        asserts.assert_equal(response.status, 3,
-                             "The response should contain a ChangeToModeResponse command "
-                             "with the Status set to InvalidInMode(0x03).")
+        self.print_step("7a", "Read FeatureMap Attribute")
+        directmodech = await self.read_feature_map_attribute()
+        directmode_enabled = directmodech & self.directmodech_bit_map
```
[the remark this answers is not in the document]
Comment (Harshith-GRL) [reply src/python_testing/TC_RVCCLEANM_2_2.py]: Re-named variables as per comments

File: src/python_testing/TC_RVCCLEANM_2_2.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Harshith-GRL) [reply src/python_testing/TC_RVCCLEANM_2_2.py]: Re-named variables as per comments

File: src/python_testing/TC_RVCRUNM_2_2.py [hunk 8]
```diff
@@ -122,6 +131,9 @@ async def test_TC_RVCRUNM_2_2(self):
                          "PIXIT.RVCRUNM.MODE_A:<mode id> \n"
                          "PIXIT.RVCRUNM.MODE_B:<mode id>")
 
+        # TODO Replace 0x8000 with python object of RVCRUN FEATURE bit when implemented
+        # 0x8000 corresponds to 16 bit DIRECTMODECH Feature map
+        self.directmodech_bit_map = 0x8000
```
[the remark this answers is not in the document]
Comment (Harshith-GRL) [reply src/python_testing/TC_RVCRUNM_2_2.py]: Re-named variables as per comments

File: src/python_testing/TC_RVCCLEANM_2_2.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Harshith-GRL) [reply src/python_testing/TC_RVCCLEANM_2_2.py]: Statements are moved before  "if" condition to remove duplicate statements

**Comment by rbultman** [comment]
Am I mistaken or did a lot of new stuff get added to this PR since yesterday that seemingly has nothing to do with the DirectModeChange feature?

**Comment by kiel-apple** [comment]
> Am I mistaken or did a lot of new stuff get added to this PR since yesterday that seemingly has nothing to do with the DirectModeChange feature?

faulty merge is my guess

**Review by rbultman — APPROVED** [review]

**Pull request merged**